app/src/main/java/com/keylesspalace/tusky/util/LinkHelper.kt

```diff
@@ -0,0 +1,264 @@
+
+import java.util.HashSet
+
+/**
+ * Finds links, mentions, and hashtags in a piece of text and makes them clickable, associating
+ * them with callbacks to notify when they're clicked.
+ *
+ * @param view the returned text will be put in
+ * @param content containing text with mentions, links, or hashtags
+ * @param mentions any '@' mentions which are known to be in the content
+ * @param listener to notify about particular spans that are clicked
+ */
+fun setClickableText(view: TextView, content: Spanned, mentions: Array<Status.Mention>?, listener: LinkListener) {
+    val builder = SpannableStringBuilder(content)
+    highlightSpans(builder, view.linkTextColors.defaultColor)
+    val urlSpans = builder.getSpans(0, content.length, URLSpan::class.java)
+    for (span in urlSpans) {
+        replaceSpan(builder, span, getLinkSpan(span.url, listener), builder.getSpanStart(span),
+                builder.getSpanEnd(span), builder.getSpanFlags(span))
+    }
+
+    val otherSpans = builder.getSpans(0, content.length, ForegroundColorSpan::class.java)
+    val usedMentionIds = HashSet<String>()
+
+    for (span in otherSpans) {
+        val start = builder.getSpanStart(span)
+        val end = builder.getSpanEnd(span)
+        val flags = builder.getSpanFlags(span)
+        val text = builder.subSequence(start, end)
+
+        val customSpan = when (text[0]) {
+            '#' -> getTagSpan(text.substring(1), listener)
+            '@' -> {
+                if (mentions != null && mentions.isNotEmpty()) {
+                    /* There may be multiple matches for users on different instances with the same
+                     * username. If a match has the same domain we know it's for sure the same, but if
+                     * that can't be found then just go with whichever one matched last. */
+                    val id: String? = firstUnusedMention(text.substring(1), mentions, usedMentionIds)
+                    if (id != null) {
+                        usedMentionIds.add(id)
+                        getAccountSpan(id, listener)
+                    } else {
+                        null
+                    }
+                } else {
+                    null
+                }
+            }
+            else -> null
+        }
+
+        replaceSpan(builder, span, customSpan, start, end, flags)
+    }
+
+    view.text = builder
+    view.linksClickable = true
+    view.movementMethod = LinkMovementMethod.getInstance()
+}
+
+/**
+ * Replace a span within a spannable string builder
+ * @param builder the builder to replace spans within
+ * @param oldSpan the span to be replaced
+ * @param newSpan the new span to be used
+ * @param start the start of the new span
+ * @param end the end of the new span
+ * @param flags the flags to be used for the new span
+ */
+private fun replaceSpan(builder: SpannableStringBuilder, oldSpan: CharacterStyle?, newSpan: ClickableSpan?, start: Int, end: Int, flags: Int) {
```

Comment:
You're always getting start and end from the old span. Maybe it makes sense to remove these arguments? Is there a scenario where they don't match?